File: implementations/python-fastapi/app/currencies.py
"""Supported ISO 4217 currency codes (the 49 pinned by the contract)."""

from __future__ import annotations

SUPPORTED_CURRENCIES: frozenset[str] = frozenset(
    {
        # Major
        "USD", "EUR", "JPY", "GBP", "CHF", "CAD", "AUD", "NZD",
        # European
        "SEK", "NOK", "DKK", "CZK", "PLN", "HUF", "RON", "BGN", "ISK",
        # Asian
        "CNY", "HKD", "SGD", "KRW", "INR", "THB", "MYR", "IDR", "PHP", "TWD", "VND",
        # Latin American
        "BRL", "MXN", "ARS", "CLP", "COP", "PEN",
        # Middle Eastern
        "AED", "SAR", "ILS", "QAR", "KWD", "BHD",
        # Other
        "ZAR", "TRY", "RUB", "UAH", "EGP", "NGN", "KES", "MAD", "PKR",
    }
)


def is_supported(code: str | None) -> bool:
    """Case-insensitive membership test against the supported set."""
    return bool(code) and code.upper() in SUPPORTED_CURRENCIES

File: implementations/ts-nestjs/src/common/currencies.ts
/**
 * Supported ISO 4217 alphabetic currency codes (49 codes).
 *
 * Currency is required and validated against this set only when `price > 0`;
 * it is ignored/optional for free articles (`price == 0`).
 */
export const SUPPORTED_CURRENCIES: string[] = [
  // Major currencies
  'USD', 'EUR', 'JPY', 'GBP', 'CHF', 'CAD', 'AUD', 'NZD',
  // European currencies
  'SEK', 'NOK', 'DKK', 'CZK', 'PLN', 'HUF', 'RON', 'BGN', 'ISK',
  // Asian currencies
  'CNY', 'HKD', 'SGD', 'KRW', 'INR', 'THB', 'MYR', 'IDR', 'PHP', 'TWD', 'VND',
  // Latin American currencies
  'BRL', 'MXN', 'ARS', 'CLP', 'COP', 'PEN',
  // Middle Eastern currencies
  'AED', 'SAR', 'ILS', 'QAR', 'KWD', 'BHD',
  // Other currencies
  'ZAR', 'TRY', 'RUB', 'UAH', 'EGP', 'NGN', 'KES', 'MAD', 'PKR',
];

const SUPPORTED_SET = new Set(SUPPORTED_CURRENCIES);

export function isSupportedCurrency(code: string | null | undefined): boolean {
  return typeof code === 'string' && SUPPORTED_SET.has(code);
}
